### PR TITLE
RxSwift to Combine translation

### PR DESCRIPTION
**Pull request by alejandro-ruiz-ts**
### PR's key points
RxSwift to Combine translation of the FLUX architecture contained in Mini. The goal is to implement a Mini with the native reactive iOS library called Combine that can be used in a native reactive application with SwiftUI (or even with UIKit). To do this, these changes have been made to the following Mini core files:
- `Action.swift`: the Action protocol remains the same, it does not depend on RxSwift.
- `ActionReducer.swift`: the Reducer class will now inherit from `Cancellable` instead of `Disposable`. Its dispose() method is replaced by the cancel() method.
- `Dispatcher.swift`: the `DispatcherSubscription` class now inherits from `Cancellable` instead of `Disposable`. Its `dispose()` method is replaced by the `cancel()` method.
- `Middleware.swift`: the `Middleware` logic remains the same, it does not depend on RxSwift.
- `ReducerGroup.swift`: the `GroupReducer` protocol and its variable `disposeBag` will now be of type `Cancellable`. The disposeBag will be an array of `Cancellables`. The `dispose()` function is replaced by the `cancel()` function. 
- `Service.swift`: the `Service` protocol remains the same, it does not depend on RxSwift.
- `State.swift`: the `StateType` protocol is now a simple basic protocol that allows us to differentiate and group the `State` classes of our application, nothing more. In Combine the `State` must be `classes` unlike before, which were `Structs`. This is so because only classes can inherit from `ObservableObject`, whose functionality thanks to Combine will now allow us to listen to state changes.
- `Store.swift`: this class now consists of a `State` of type `ObservableObject` and a `StoreController` of type `Cancellable`. The `ObservableObject` of Combine will be the one that will allow us to subscribe to the changes that occur in that state. This allows us to get rid of all the extensions and functions dedicated to listen to the states that we had when we depended on RxSwift. In the same way, `Cancellable` will be the new `Disposable` of our application. The code of the class has been quite reduced and clear.

Regarding the rest of Mini's content: 
- **Utils**: prematurely, during the study of the original Mini to try to understand and know how everything works, we started by translating the operator extensions contained in `ObservableType+Extensions.swift` from RxSwift to Combine. Although they are not ultimately needed within a project that uses Combine+SwiftUI, the native library provides the same operators. All other classes in this directory remain the same, they do not depend on RxSwift.
- **MiniTasks & MiniPromises**: these two modules are still pending to be translated into Combine, specifically their `PrimitiveSequenceType` extension. They are not strictly necessary for the architecture of a SwiftUI+Combine project. The code has been commented and marked as TODO.
- **TestMiddleware**: remains the same as it does not depend on RxSwift.
- **LoggingService**: remains the same as it does not depend on RxSwift.

Regarding the `package.swift` file for package configuration:
- Dependencies and targets that referenced the RxSwift and RxOptional libraries have been removed.

This new MiniCombine has been tested and verified in a project with FLUX architecture using the Assembler and dispatching actions to the stores. Its performance is correct and similar to what we had with MiniRxSwift.

 
### How to review this PR?
…

### CODE REVIEW

File: Sources/Mini/ActionReducer.swift [hunk 1]
```diff
@@ -1,35 +1,19 @@
-/*
- Copyright [2019] [BQ]
```
Comment (minuscorp) [inline Sources/Mini/ActionReducer.swift:2]: Keep the copyright and the license in every file as it was before. Update the license holder to HyperDevs since 2021.

File: Sources/Mini/ActionReducer.swift [hunk 2]
```diff
@@ -1,35 +1,19 @@
...
 import Foundation
-import RxSwift
-
+import Combine
 /**
  The `Reducer` defines the behavior to be executed when a certain
  `Action` object is received.
  */
-public class Reducer<A: Action>: Disposable {
+ @available(iOS 13.0, *)
```
Comment (minuscorp) [inline Sources/Mini/ActionReducer.swift:7]: Remove call `@available` annotations and instead bump the minimum target version in the `Package.swift` file.

File: Sources/Mini/Dispatcher.swift [hunk 3]
```diff
@@ -133,15 +127,29 @@ public final class Dispatcher {
             if let action = object as? T {
                 completion(action)
             } else {
-                fatalError("Casting to \(tag) failed")
+                print("MiniError: Casting to \(tag) failed")
```
Comment (minuscorp) [inline Sources/Mini/Dispatcher.swift:130]: This must crash because data revving has been broken, if this happens, something about threading is wrong.

File: Sources/Mini/Dispatcher.swift [hunk 4]
```diff
@@ -100,6 +85,15 @@ public final class Dispatcher {
         )
         return registerInternal(subscription: subscription)
     }
+    
+    private func getNewSubscriptionId() -> Int {
```
Comment (minuscorp) [inline Sources/Mini/Dispatcher.swift:89]: This method is not thread safe. Data races may occur and crash. Switch back to atomic operations if possible.

File: Sources/Mini/ReducerGroup.swift [hunk 5]
```diff
@@ -1,35 +1,22 @@
...
- */
-
 import Foundation
-import RxSwift
+import Combine
 
-public protocol Group: Disposable {
-    var disposeBag: CompositeDisposable { get }
+@available(iOS 13.0, *)
+public protocol GroupReducer: Cancellable {
```
Comment (minuscorp) [inline Sources/Mini/ReducerGroup.swift:5]: Can be renamed to `CancellableGroup` for clarification.

File: Sources/Mini/StateType.swift [hunk 6]
```diff
@@ -1,28 +0,0 @@
...
- Unless required by applicable law or agreed to in writing, software
- distributed under the License is distributed on an "AS IS" BASIS,
- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- See the License for the specific language governing permissions and
- limitations under the License.
- */
-
-import Foundation
-
-public protocol StateType {
```
Comment (minuscorp) [inline Sources/Mini/StateType.swift:19]: Why deleting this protocol?

File: Sources/Mini/Store.swift [hunk 7]
```diff
@@ -1,138 +1,30 @@
...
-        }
-        set {
-            queue.sync {
-                if !newValue.isEqual(to: _state) {
-                    _state = newValue
-                    objectWillChange.onNext(state)
-                }
-            }
+        didSet {
+            objectWillChange.send()
```
Comment (minuscorp) [inline Sources/Mini/Store.swift:14]: Only different states are propagated through the architecture, why the change?

File: Sources/Mini/Store.swift [hunk 8]
```diff
@@ -1,138 +1,30 @@
...
-     */
-    public var reducerGroup: ReducerGroup {
-        return ReducerGroup()
-    }
-}
-
-public class Store<State: StateType, StoreController: Disposable>: ObservableType, StoreType {
-    public typealias Element = State
+@available(iOS 13.0, *)
+public class Store<State: ObservableObject, StoreController: Cancellable>: ObservableObject {
```
Comment (minuscorp) [inline Sources/Mini/Store.swift:5]: `State` must conform to `StateType`, if it has to conform to `ObservableObject`, delegate it to the former protocol.

Comment (minuscorp) [inline Sources/Mini/Store.swift:5]: Also `StoreType` must be confirmed in the `Store` class, changes needed must be performed through the protocol.

File: Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ limitations under the License.
+ */
+
+import Foundation
+import Combine
+
+@available(iOS 13.0, *)
+public extension Publisher {
+
+    func filterKey(_ keyPath: KeyPath<Output, Bool>) -> AnyPublisher<Output, Failure> {
```
Comment (minuscorp) [inline Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift:23]: `Publishers.Filter<Self>`?

**Review by minuscorp — CHANGES_REQUESTED** [review]
Some comments, this can be indeed used from the branch but not ready for merging, there's functionality left to be translated to bridge with previous' library versions.

File: Sources/Mini/ActionReducer.swift (hunk 1 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/ActionReducer.swift:2]: Fixed.

Comment (alejandro-ruiz-ts) [reply Sources/Mini/ActionReducer.swift:2]: Fixed!

File: Sources/Mini/ActionReducer.swift (hunk 2 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/ActionReducer.swift]: Fixed, now target version is 13.0.

File: Sources/Mini/Dispatcher.swift (hunk 3 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Dispatcher.swift]: Ok, replaced. I like to control errors rather than cause crashes, but it doesn't seem to be a frequent error case.

File: Sources/Mini/Dispatcher.swift (hunk 4 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Dispatcher.swift:103]: Ok, I have added an Atomic class to control this value in a safe way. It should be similar to the way you had it implemented before.

File: Sources/Mini/StateType.swift (hunk 6 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/StateType.swift:19]: I have not seen it necessary at any point in the flow, in fact I could delete it completely but I leave it in case in Rx it makes more sense.

File: Sources/Mini/Store.swift (hunk 7 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Store.swift]: Now state is an @Published variable that allows us to listen when the whole state changes or when any of its parameters change individually. This works quite well for us from Combine, it is much simpler than before.

File: Sources/Mini/Store.swift (hunk 8 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Store.swift:21]: I do not understand the need for the StateType and its implication in these classes. Right now in our flow it seems irrelevant. I would appreciate if you could tell me a bit more about this and see if I understand why to include it here.

Comment (alejandro-ruiz-ts) [reply Sources/Mini/Store.swift:21]: I do not understand the need for the StateType and its implication in these classes. Right now in our flow it seems irrelevant. I would appreciate if you could tell me a bit more about this and see if I understand why to include it here.

File: Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift (hunk 9 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift:23]: This extension of operators I have been very close to delete it all, since in Combine none of them are needed, it has its own native operators. But I understand that for use in Rx it may still make sense to keep them, but they are irrelevant for us, they would not be needed. If you want me to adapt or change any of this, let me know exactly what.

File: Sources/Mini/Store.swift (hunk 8 above)
Comment (minuscorp) [reply Sources/Mini/Store.swift:21]: Because working with protocols allow extensions to be done into generics while classes don't. It's a matter of extensibility.

File: Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift (hunk 9 above)
Comment (minuscorp) [reply Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift:23]: The return type of the function, wouldn't be more correct to be my suggestion?

File: Package.resolved [hunk 10]
```diff
@@ -77,8 +77,8 @@
         "package": "Nimble",
         "repositoryURL": "https://github.com/Quick/Nimble",
         "state": {
-          "branch": "master",
-          "revision": "eea5843b34beb559dd51cf004953f75028e47add",
+          "branch": "main",
```
Comment (minuscorp) [inline Package.resolved:80]: Let's lock into a version `.upToNextMajor`

File: Sources/Mini/Dispatcher.swift [hunk 11]
```diff
@@ -230,3 +231,26 @@ public final class DispatcherSubscription: Comparable, Disposable {
         return lhs.priority <= rhs.priority
     }
 }
+
+public class AtomicCounter {
```
Comment (minuscorp) [inline Sources/Mini/Dispatcher.swift:235]: This is something internal to the library, shouldn't be exposed as public

File: Sources/Mini/Dispatcher.swift [hunk 12]
```diff
@@ -230,3 +231,26 @@ public final class DispatcherSubscription: Comparable, Disposable {
...
+
+    private var mutex = pthread_mutex_t()
```
Comment (minuscorp) [inline Sources/Mini/Dispatcher.swift:237]: `os_unfair_lock_lock` is the fastest

**Review by minuscorp — CHANGES_REQUESTED** [review]
Still, we cannot ensure that this library will be used only in SwiftUI, but in UIKit, so there're still things to do and translate to keep the old implementation removing just RxSwift dependency.

File: Sources/Mini/Store.swift [hunk 13]
```diff
@@ -1,138 +1,42 @@
...
-
-    var state: State { get set }
-    var dispatcher: Dispatcher { get }
-    var reducerGroup: ReducerGroup { get }
-
-    func replayOnce()
-}
-
-extension StoreType {
-    /**
```
[the remark this answers is not in the document]
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Store.swift:32]: Ok! Fixed.

File: Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            } else {
+                seen.append(incoming)
+                return true
+            }
+        }
+    }
+}
+
+ 
+public extension Publisher where Self.Output: OptionalType {
```
[the remark this answers is not in the document]
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift:55]: First filter changed but second comment, about extension, causes this function to stop working:

```
    func select<T: OptionalType>(_ keyPath: KeyPath<Output, T>) -> AnyPublisher<T.Wrapped, Self.Failure> where T.Wrapped: Equatable {
        map(keyPath)
            .filterNil()
            .removeDuplicates()
            .eraseToAnyPublisher()
    }
```
This error: `Referencing instance method 'filterNil()' on 'Publisher' requires the types 'Publishers.MapKeyPath<Self, T>.Failure' (aka 'Self.Failure') and 'Never' be equivalent`

File: Package.resolved (hunk 10 above)
Comment (alejandro-ruiz-ts) [reply Package.resolved:80]: I have tried it but it gives me error because it does not find the library: `NIOConcurrencyHelpers`. I have left it in the main branch in the meantime.

File: Sources/Mini/Dispatcher.swift (hunk 11 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Dispatcher.swift]: Fixed!

File: Sources/Mini/Dispatcher.swift (hunk 12 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Dispatcher.swift]: Ok, changed with this approach.

File: Package.swift [hunk 15]
```diff
@@ -6,7 +6,7 @@ import PackageDescription
 let package = Package(
     name: "Mini",
     platforms: [
-        .iOS(.v11),
+        .iOS("13.0"),
```
[the remark this answers is not in the document]
Comment (alejandro-ruiz-ts) [reply Package.swift]: Updated to 5.3.

File: Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift (hunk 9 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift]: Fixed!

File: Sources/Mini/Store.swift (hunk 8 above)
Comment (alejandro-ruiz-ts) [reply Sources/Mini/Store.swift]: Ok, protocol restored and applied at both State and Store.

Comment (alejandro-ruiz-ts) [reply Sources/Mini/Store.swift]: Ok, protocol restored and applied at both State and Store.

File: Package.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (minuscorp) [reply Package.swift:9]: But you're not using atomic integers anymore don't you?

File: Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (minuscorp) [reply Sources/Mini/Utils/Extensions/ObservableType+Extensions.swift:55]: I think you need to set Failure == Never on the extension

File: Package.swift [hunk 16]
```diff
@@ -56,7 +54,7 @@ let package = Package(
         // Targets can depend on other targets in this package, and on products in packages which this package depends on.
         .target(
             name: "Mini",
-            dependencies: ["RxSwift", "NIOConcurrencyHelpers", "RxOptional"]
+            dependencies: ["NIOConcurrencyHelpers"]
```
Comment (minuscorp) [inline Package.swift:57]:
```suggestion
            dependencies: [.product(package: "swift-nio", package: "NIOConcurrencyHelpers"]
```

Comment (minuscorp) [reply Package.swift:57]: But again, if you're not using atomics anymore from swift-nio, you can delete the dependency

**Review by minuscorp — CHANGES_REQUESTED** [review]
